### DISCUSSION TITLE
Is it possible to make atom-level Morgan fingerprints? - Help Request for a project

### DISCUSSION THREAD

**Discussion by wan-nie** [rdkit/rdkit#5390] [Q&A]
### Background
Recently, I am preparing a manuscript about the application of graph neural network (GNN) in the molecular properties prediction task. One reviewer gives the followed advice.
> I suggest to build ordinary 2D models based on count-based Morgan fingerprints of radius 2 and 3 (not binary ones) and models based on graph convolution approach.

In terms of preparing the input of GNN, each molecule must be encoded to a `N x K` matrix (for node information only), where `N` is the number of atoms in the molecule and `K` is the dimension of the features. That is, the encoded input is an **atom (node)-level** matrix.  As I understand it, Morgan fingerprints encode which substructures are present in a molecule, and count-based vectors track the number of times each substructure appears. Therefore, the count-based Morgan fingerprints are molecule-level vectors with `1 x K` shape which cannot be used in the GNN model.

### Questions
- Is my understanding of the reviewer's advice correct?
- Is it possible to make atom-level Morgan fingerprints?

Any suggestions are appreciated.

**Comment by greglandrum** [comment]
@wan-nie you are correct that it is possible that a morgan fingerprint is normally `1xK`. 
I can come up with two possible ways to address the reviewer's concerns, and maybe it's worth doing both:
1. Build a simpler model using the Morgan fingerprints and either a random forest or XGBoost and compare the results from that with your GNN. If I were the reviewer I would want to see this comparison (I always want to see the "state of the art" methods compared to the simple things that we've been using for years).
2.  You can get an `NxK` matrix by finding which atoms are at the center of each bit. A description of how to do this is here: https://www.rdkit.org/docs/GettingStartedInPython.html#explaining-bits-from-morgan-fingerprints

**Reply by wan-nie** [reply]
Dear @greglandrum , thanks for the answer. I am glad to receive your informative advice,. The first suggestion is easy to understand and certainly worth doing. 

For the second one, my initial thought was to make the `N x K` matrix to indicate whether the atom (rows) is at the center of each bit (columns). That is, first, the `K` bit ids are sorted ascendingly.  Second, if the `n`th atom is at the `k`th bit center, then set `matrix[n,k] = 1`. And the rest of the matrix is set to zero. 

However, the thing is, different molecules may have a different number of bit ids(substructures), which should be avoided. Therefore, what I have come up with is to iterate all molecules in the data set to determine all possible bit ids `bit_ids`. For each molecule, `K` is set to the number of all possible bit ids `feature_dim = len(bit_ids)` so that all molecules can have consistent feature dimensions. 

I do not know if I understand your second suggestion correctly.  Here is a sample code.  Any suggestions are appreciated. 


```python
def change_info_to_matrix(atom_num, feature_dim, info):
    out_matrix = np.zeros((atom_num, feature_dim))
    for bit_id, atom_ids in info.items():
        out_matrix[atom_ids, bit_id] = 1
    return out_matrix


def get_morgan_matrix(smiles_ls, radius):
    bit_id_ls = []
    info_ls = []
    atom_num_ls = []
    for smiles in smiles_ls:
        info = {}
        m = Chem.MolFromSmiles(smiles)
        m = Chem.AddHs(m)  # including hydrogen
        atom_num = len(m.GetAtoms())
        fp = AllChem.GetMorganFingerprint(m, radius, bitInfo=info)

        bit_id_ls += list(info.keys())
        info_ls.append(info)
        atom_num_ls.append(atom_num)

    # create new_info_ls
    bit_ids = np.sort(np.unique(bit_id_ls))
    new_info_ls = []
    for info in info_ls:
        new_info = {}
        for bit_id, atom_id_radius_tuple in info.items():
            new_bit_id = np.argmax(bit_ids == bit_id)  # the idx in bit_ids
            atom_ids = [atom_id for atom_id, radius in atom_id_radius_tuple]
            new_info[new_bit_id] = atom_ids
        new_info_ls.append(new_info)

    # convert new_info_ls to matrix ls
    matrix_ls = []
    feature_dim = len(bit_ids)
    for atom_num, new_info in zip(atom_num_ls, new_info_ls):
        matrix_ls.append(change_info_to_matrix(atom_num, feature_dim, new_info))
```
